Document/0x05c-Reverse-Engineering-and-Tampering.md

```diff
@@ -1497,7 +1488,368 @@ The hooked function outputted the decrypted string. You extracted the secret str
+0x7d39d2024d hit0_9 Hello;
+0x7d3aa4d274 hit0_10 Hello
+```
+
+Now you'd like to know where are these addresses actually. You may do so by running the `\dm.` command for all `@@` hits matching the glob `hit0_*`:
+
+```bash
+[0x00000000]> \dm.@@ hit0_*
+0x0000000013100000 - 0x0000000013140000 rw- /dev/ashmem/dalvik-main space (region space) (deleted)
+0x0000000013100000 - 0x0000000013140000 rw- /dev/ashmem/dalvik-main space (region space) (deleted)
+0x0000000013100000 - 0x0000000013140000 rw- /dev/ashmem/dalvik-main space (region space) (deleted)
+0x00000000703c2000 - 0x00000000709b5000 rw- /data/dalvik-cache/arm64/system@framework@boot-framework.art
+0x0000007d1c499000 - 0x0000007d1c49a000 r-x /data/app/sg.vantagepoint.helloworldjni-1/lib/arm64/libnative-lib.so
+0x0000007d1c4a9000 - 0x0000007d1c4aa000 r-- /data/app/sg.vantagepoint.helloworldjni-1/lib/arm64/libnative-lib.so
+0x0000007d1c516000 - 0x0000007d1c54d000 r-- /data/app/sg.vantagepoint.helloworldjni-1/base.apk
+0x0000007d30a00000 - 0x0000007d30c00000 rw-
+0x0000007d396bc000 - 0x0000007d3a998000 r-- /system/framework/arm64/boot-framework.vdex
+0x0000007d396bc000 - 0x0000007d3a998000 r-- /system/framework/arm64/boot-framework.vdex
+0x0000007d3a998000 - 0x0000007d3aa9c000 r-- /system/framework/arm64/boot-ext.vdex
+```
+
+Additionally, you can search for occurrences of the [wide version of the string](https://en.wikipedia.org/wiki/Wide_character "Wide character") (`\/w`) and, again, check their memory regions:
+
+```bash
+[0x00000000]> \/w Hello
+Searching 10 bytes: 48 00 65 00 6c 00 6c 00 6f 00
+hits: 6
+0x13102acc hit1_0 480065006c006c006f00
+0x13102b9c hit1_1 480065006c006c006f00
+0x7d30a53aa0 hit1_2 480065006c006c006f00
+0x7d30a872b0 hit1_3 480065006c006c006f00
+0x7d30bb9568 hit1_4 480065006c006c006f00
+0x7d30bb9a68 hit1_5 480065006c006c006f00
+
+[0x00000000]> \dm.@@ hit1_*
+0x0000000013100000 - 0x0000000013140000 rw- /dev/ashmem/dalvik-main space (region space) (deleted)
+0x0000000013100000 - 0x0000000013140000 rw- /dev/ashmem/dalvik-main space (region space) (deleted)
+0x0000007d30a00000 - 0x0000007d30c00000 rw-
+0x0000007d30a00000 - 0x0000007d30c00000 rw-
+0x0000007d30a00000 - 0x0000007d30c00000 rw-
+0x0000007d30a00000 - 0x0000007d30c00000 rw-
+```
+
+They are in the same rw- region as one of the previous strings (`0x0000007d30a00000`). Note that searching for the wide versions of strings is sometimes the only way to find them as you'll see in the following section.
+
+In-memory search can be very useful to quickly know if certain data is located in the main app binary, inside a shared library or in another region. You may also use it to test the behavior of the app regarding how the data is kept in memory. For instance, you could analyze an app that performs a login and search for occurrences of the user password. Also, you may check if you still can find the password in memory after the login is completed to verify if this sensitive data is wiped from memory after its use.
```

Comment:
Maybe this can be extended to explain that using keys to encrypt data in a binary can be found like this as well? (to further push people towards using the AndroidKeyStore ^^)

Reply:
nice point, I'll add something on that ;)

Reply:
done